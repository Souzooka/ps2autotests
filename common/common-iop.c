#include <stdio.h>
#include <stdarg.h>
#include <xprintf.h>

#define SCHEDF_BUFFER_SIZE 65536

char schedfBuffer[65536];
unsigned int schedfBufferPos = 0;

void schedf(const char *format, ...) {
	va_list args;
	va_start(args, format);
	schedfBufferPos += vsnprintf(schedfBuffer + schedfBufferPos, SCHEDF_BUFFER_SIZE, format, args);
	va_end(args);
}

void flushschedf() {
	printf("%s", schedfBuffer);
	schedfBuffer[0] = '\0';
	schedfBufferPos = 0;
}
