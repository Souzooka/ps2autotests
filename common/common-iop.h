#ifndef PS2AUTOTESTS_COMMON_IOP_H
#define PS2AUTOTESTS_COMMON_IOP_H

// Defines for MSVC highlighting.  Not intended to actually compile with msc.
#ifdef _MSC_VER
#define __STDC__
#define _IOP
#define __attribute__(x)
#endif

#include <stdio.h>
#include <tamtypes.h>

void schedf(const char *format, ...);
void flushschedf();

#endif
