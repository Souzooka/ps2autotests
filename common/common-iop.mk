# _____     ___ ____     ___ ____
#  ____|   |    ____|   |        | |____|
# |     ___|   |____ ___|    ____| |    \    PS2DEV Open Source Project.
#-----------------------------------------------------------------------
# Copyright 2001-2004, ps2dev - http://www.ps2dev.org
# Licenced under Academic Free License version 2.0
# Review ps2sdk README & LICENSE files for further details.
#
# $Id: Makefile.eeglobal_cpp_sample 1525 2009-01-23 00:24:13Z radad $

# Define pefixes for the various toolchain binaries.
include $(COMMON_DIR)/common-defines.mk

ASFLAGS_TARGET = -mcpu=r3000

CFLAGS_TARGET  = -miop
ASFLAGS_TARGET = -march=r3000
LDFLAGS_TARGET = -miop

# include dir
IOP_INCS := -I$(PS2SDK)/iop/include -I$(PS2SDK)/common/include -I$(COMMON_DIR) \
	-I. $(IOP_INCS)
# C compiler flags
IOP_CFLAGS := $(CFLAGS_TARGET) -O2 -G0 $(IOP_INCS) $(IOP_CFLAGS)

# linker flags
IOP_LDFLAGS := $(LDFLAGS_TARGET) -nostdlib -L$(PS2SDK)/iop/lib $(IOP_LDFLAGS)

# asssembler flags
IOP_ASFLAGS := $(ASFLAGS_TARGET) -EL -G0 $(IOP_ASFLAGS)

# link with following libraries (libs need to be defined multiple times in order for linking to work!!)
IOP_LIBS += -lkernel -lgcc

# Externally defined variables: IOP_BIN, IOP_OBJS, IOP_LIB

%.o : %.c
	$(IOP_CC) $(IOP_CFLAGS) -c $< -o $@

%.o : %.S
	$(IOP_CC) $(IOP_CFLAGS) -c $< -o $@

%.o : %.s
	$(IOP_AS) $(IOP_ASFLAGS) $< -o $@

%.o : %.lst
	echo "#include \"irx_imports.h\"" > build-imports.c
	cat $< >> build-imports.c
	$(IOP_CC) $(IOP_CFLAGS) -c build-imports.c -o $@
	rm -f build-imports.c

%.irx : %.o $(EXTRA_OBJS)
	$(IOP_CC) $(IOP_LDFLAGS) -o $@ $< $(EXTRA_OBJS) $(IOP_LIBS)

all: $(TARGETS:=.irx)

clean:
	-$(RM) -f $(TARGETS:=.irx) $(TARGETS:=.o)

rebuild: clean all
