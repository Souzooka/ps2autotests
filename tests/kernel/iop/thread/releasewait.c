#include <common-iop.h>
#include <thbase.h>
#include <thsemap.h>
#include <thevent.h>
#include <thmsgbx.h>
#include "thread-common.h"

#define STACK_SIZE 0x800

void nullThreadProc(u32 param) {
	//Do nothing
}

void sleepThreadProc(s32 *result) {
	*result = SleepThread();
}

void delayThreadProc(s32 *result) {
	*result = DelayThread(~0);
}

void semaThreadProc(s32 *result) {
	iop_sema_t dummySemaInfo = {};
	dummySemaInfo.attr    = 0;
	dummySemaInfo.option  = 0;
	dummySemaInfo.initial = 0;
	dummySemaInfo.max     = 1;
	u32 dummySema = CreateSema(&dummySemaInfo);
	*result = WaitSema(dummySema);
}

void eventFlagThreadProc(s32 *result) {
	iop_event_t dummyEventFlagInfo = {};
	dummyEventFlagInfo.attr = 0;
	dummyEventFlagInfo.bits = 0;
	u32 dummyEventFlag = CreateEventFlag(&dummyEventFlagInfo);
	*result = WaitEventFlag(dummyEventFlag, ~0, WEF_CLEAR | WEF_AND, NULL);
}

void messageBoxThreadProc(s32 *result) {
	iop_mbx_t dummyMessageBoxInfo = {};
	dummyMessageBoxInfo.attr = 0;
	dummyMessageBoxInfo.option = 0;
	u32 dummyMessageBox = CreateMbx(&dummyMessageBoxInfo);
	*result = ReceiveMbx(NULL, dummyMessageBox);
}

void doReleaseWaitThread(s32 threadId) {
	s32 result = ReleaseWaitThread(threadId);
	if(result >= 0) {
		schedf("succeeded -> result: %d\n", result);
	} else {
		schedf("failed -> result: %d\n", result);
	}
}

void testRunningThread() {
	schedf("running thread: ");
	
	s32 currentThreadPriority = getThreadPriority(TH_SELF);
	s32 threadId = createTestThread(&nullThreadProc, currentThreadPriority + 1, STACK_SIZE);
	StartThread(threadId, 0);
	
	doReleaseWaitThread(threadId);
	
	//Let the thread execute (terminate doesn't seem to work properly otherwise)
	DelayThread(1);
	
	TerminateThread(threadId);
	DeleteThread(threadId);
	
	flushschedf();
}

void testDormantThread() {
	schedf("dormant thread: ");
	
	s32 threadId = createTestThread(&nullThreadProc, 0x20, STACK_SIZE);
	
	doReleaseWaitThread(threadId);
	
	DeleteThread(threadId);
	
	flushschedf();
}

void testInvalidThread() {
	schedf("invalid thread: ");
	
	doReleaseWaitThread(~0);
	
	flushschedf();
}

void testWaitingThread(const char *waitType, void *threadProc) {
	schedf("%s thread:\n", waitType);
	
	s32 waitResult = 0;
	s32 currentThreadPriority = getThreadPriority(TH_SELF);
	s32 threadId = createTestThread(threadProc, currentThreadPriority - 1, STACK_SIZE);
	StartThread(threadId, &waitResult);
	
	schedf("  ReleaseWaitThread returned: ");
	doReleaseWaitThread(threadId);
	
	schedf("  wait function returned: %d\n", waitResult);
	
	TerminateThread(threadId);
	DeleteThread(threadId);
	
	flushschedf();
}

void testSelf() {
	schedf("self thread:\n");
	
	schedf("  With TH_SELF: ");
	doReleaseWaitThread(TH_SELF);
	
	schedf("  With current thread id: ");
	doReleaseWaitThread(GetThreadId());
	
	flushschedf();
}

int _start(int argc, char **argv) {
	printf("-- TEST BEGIN\n");
	
	testRunningThread();
	testDormantThread();
	testInvalidThread();
	testWaitingThread("sleeping", &sleepThreadProc);
	testWaitingThread("delayed", &delayThreadProc);
	testWaitingThread("waiting semaphore", &semaThreadProc);
	testWaitingThread("waiting event flag", &eventFlagThreadProc);
	testWaitingThread("waiting message box", &messageBoxThreadProc);
	testSelf();
	
	printf("-- TEST END\n");
	return 0;
}
