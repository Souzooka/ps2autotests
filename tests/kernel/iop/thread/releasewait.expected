-- TEST BEGIN
running thread: failed -> result: -416
dormant thread: failed -> result: -416
invalid thread: failed -> result: -407
sleeping thread:
  ReleaseWaitThread returned: succeeded -> result: 0
  wait function returned: -418
delayed thread:
  ReleaseWaitThread returned: succeeded -> result: 0
  wait function returned: -418
waiting semaphore thread:
  ReleaseWaitThread returned: succeeded -> result: 0
  wait function returned: -418
waiting event flag thread:
  ReleaseWaitThread returned: succeeded -> result: 0
  wait function returned: -418
waiting message box thread:
  ReleaseWaitThread returned: succeeded -> result: 0
  wait function returned: -418
self thread:
  With TH_SELF: failed -> result: -406
  With current thread id: failed -> result: -406
-- TEST END
